captum/attr/_core/kernel_shap.py

```diff
@@ -0,0 +1,324 @@
+        # end-point criteria. In practice, it is sufficient to
+        # make this weight substantially larger so setting this
+        # weight to 100 (all other weights are < 1).
+        similarities = 100.0
+    return torch.tensor([similarities])
+
+
+class KernelShap(Lime):
+    r"""
+    Kernel SHAP is a method that uses the LIME framework to compute
+    Shapley Values. Setting the loss function, weighting kernel and
+    regularization terms appropriately in the LIME framework allows
+    theoretically obtaining Shapley Values more efficiently than
+    directly computing Shapley Values.
+
+    Note that KernelShap requires having sklearn installed, since
+    sklearn is used to train the surrogate linear model.
+
+    More information regarding this method and proof of equivalence
+    can be found in the original paper here:
+    https://arxiv.org/abs/1705.07874
+    """
+
+    def __init__(self, forward_func: Callable) -> None:
+        r"""
+        Args:
+
+            forward_func (callable): The forward function of the model or
+                        any modification of it
+        """
+        Lime.__init__(
+            self,
+            forward_func,
+            linear_regression_interpretable_model_trainer,
+            kernel_shap_similarity_kernel,
+        )
+
+    @log_usage()
+    def attribute(  # type: ignore
+        self,
+        inputs: TensorOrTupleOfTensorsGeneric,
+        baselines: BaselineType = None,
+        target: TargetType = None,
+        additional_forward_args: Any = None,
+        feature_mask: Union[None, Tensor, Tuple[Tensor, ...]] = None,
+        n_samples: int = 25,
+        perturbations_per_eval: int = 1,
+        return_input_shape: bool = True,
+    ) -> TensorOrTupleOfTensorsGeneric:
+        r"""
+        This method attributes the output of the model with given target index
+        (in case it is provided, otherwise it assumes that output is a
+        scalar) to the inputs of the model using the approach described above,
+        training an interpretable model based on KernelSHAP and returning a
+        representation of the interpretable model.
+
+        It is recommended to only provide a single example as input (tensors
+        with first dimension or batch size = 1). This is because LIME is generally
+        used for sample-based interpretability, training a separate interpretable
+        model to explain a model's prediction on each individual example.
+
+        Nevertheless, a batch of inputs can also be provided as inputs, similar to
+        other perturbation-based attribution methods. If forward_func returns
+        a single value per batch (e.g. loss), then an interpretable model is
+        trained with the batch output and corresponding interpretable feature
+        vector for the batch as input. If a scalar is returned per example,
+        then the interpretable feature vector for the batch is repeated and
```

Comment:
I understand but one inconsistency with the perturbation methods is that if the output isn't scalar then we return importance scores per example, isn't it? We could potentially run forward pass only if first dimension of the input is > 1. In that case if the output isn't a scalar we can call attribute for each example(currently sequentially but we can also parallelize it) and aggregate the results. We can warn the user that this will slow the response. This will be equivalent to what we will expect from other perturbation methods as well right ?

Potentially, we can have the logic of loss for IG and other gradient-based algorithms as well. We don't support it right now but it would be interesting.

Reply:
Sure, sounds good, can update to this to be more consistent with other methods.